pepr/robustness/foolbox_wrapper.py

```diff
@@ -0,0 +1,533 @@
+            l2_dist_list.append(dist_eps)
+
+        self.attack_results["raw"] = raw_list
+        self.attack_results["clipped"] = clipped_list
+        self.attack_results["is_adv"] = is_adv_list
+        self.attack_results["success_rate"] = misclass_list
+        self.attack_results["avg_l2_distance"] = l2_dist_list
+
+        # Print attack summary
+        def _list_to_formatted_string(arr):
+            string = ""
+            # Short output if needed
+            if len(arr) <= 10:
+                for item in arr:
+                    string = string + f"{round(item, 3):>10}"
+            else:
+                for i in range(5):
+                    string = string + f"{round(arr[i], 3):>10}"
+                string = string + f"{'...':>10}"
+                for i in range(-5, 0):
+                    string = string + f"{round(arr[i], 3):>10}"
+            return string
+
+        # Print every epsilon result of attack
+        def _target_model_rows():
+            string = ""
+            for tm_i in range(len(self.target_models)):
+                string = string + f"\n{f'Target Model {tm_i + 1}:':<20}"
+                string = string + _list_to_formatted_string(
+                    self.attack_results["success_rate"][tm_i]
+                )
+            return string
+
+        logger.info(
+            "Attack Summary"
+            f"\n"
+            f"\n###################### Attack Results ######################"
+            f"\n"
+            + f"\n{'Epsilons:':<20}"
+            + _list_to_formatted_string(self.attack_pars["epsilons"])
+            + _target_model_rows()
+        )
+
+    def create_attack_report(self, save_path, pdf=False):
+        """
+        Create an attack report just for the given attack instantiation.
+
+        Parameters
+        ----------
+        save_path : str
+            Path to save the tex, pdf and asset files of the attack report.
+        pdf : bool
+            If set, generate pdf out of latex file.
+        """
+
+        # Create directory structure for the attack report, including the figure
+        # directory for the figures of the results subsubsection.
+        os.makedirs(save_path + "/fig", exist_ok=True)
+
+        self.create_attack_section(save_path)
+        report.report_generator(save_path, [self.report_section], pdf)
+
+    def create_attack_section(self, save_path):
+        """
+        Create an attack section for the given attack instantiation.
+
+        Parameters
+        ----------
+        save_path : str
+            Path to save the tex, pdf and asset files of the attack report.
+        """
+        self._report_attack_configuration()
+        self._report_attack_results(save_path)
+
+    def _gen_attack_pars_rows(self, table):
+        for key in self.pars_descriptors:
+            desc = self.pars_descriptors[key]
+            value = str(self.foolbox_attack.__dict__[key])
+
+            table.add_hline()
+            table.add_row([desc, value])
```

Comment:
We should add a docstring to this function.

Reply:
This function was supposed to be inside `_report_attack_configuration` because it's only a helper function specifically for the use in `_report_attack_configuration`. I would rather move it inside than exposing it to the API ... what do you think?

Reply:
I think we can leave the function there because docstrings of private functions (leading underscore) are not included in the documentation. So it would not be part of the API then, right?

Reply:
But it remains callable. Inside the function it would be hidden which would be better, because the function is really meant to be used by `_report_attack_configuration` **only**. Calling it from somewhere else will currently result in undefined behavior.

Reply:
Yea, that's python I guess. It's not that important, only if it remains a function a docstring should be used :)

Reply:
I've moved the function inside `_report_attack_configuration` in 9eb7669